File: 200-number-of-islands/number-of-islands.cpp
class Solution {
public:
    vector<pair<int, int> > dir = {{0,1}, {0, -1}, {1, 0}, {-1, 0}};
    void bfs(int i, int j, vector<vector<char>>& grid, vector<vector<bool>>& vis){
        int m = grid.size();
        int n = grid[0].size();
        queue<pair<int, int>> q;
        q.push({i,j});
        while(!q.empty()){
            int x = q.front().first;
            int y = q.front().second;
            q.pop();     
            if (!vis[x][y]){
                vis[x][y] = true;
                for(pair<int, int> d : dir){
                    int dx = d.first;
                    int dy = d.second;
                    if(x + dx >= 0 && x + dx < m && y + dy >= 0 && y + dy < n && !vis[x + dx][y + dy] && grid[x + dx][y + dy] == '1'){
                        q.push({x + dx, y + dy});
                    }
                }
            }
            
        }
    }
    int numIslands(vector<vector<char>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        int count = 0;
        vector<vector<bool> > vis(m, vector<bool>(n,false));
        for(int i=0; i<m; i++){
            for(int j=0; j<n; j++){
                if(grid[i][j] == '1' && !vis[i][j]){
                    bfs(i, j, grid, vis);
                    count++;
                }
            }
        }
        return count;
    }
};

File: 200-number-of-islands/number-of-islands.java
class Solution {
    
    public int numIslands(char[][] grid) {
        int[][] directions = {{0,1},{1,0},{0,-1},{-1,0}};
        int m = grid.length;
        int n = grid[0].length;

        //boolean[][] visited = new boolean[m][n];

        ArrayDeque<int[]> queue = new ArrayDeque<>();
        int island = 0;
        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                if(grid[i][j] == '1'){
                    island++;
                    queue.addLast(new int[]{i,j});
                    grid[i][j] = '0';
                    while(!queue.isEmpty()){
                        int[] pos = queue.pollFirst();
                        for(int[] dir : directions){
                            int new_x = pos[0] + dir[0]; 
                            int new_y = pos[1] + dir[1]; 
                            if(!(new_x < 0 || new_x >=m || new_y < 0 || new_y >= n) && grid[new_x][new_y] == '1'){
                                queue.addLast(new int[]{new_x, new_y});
                                grid[new_x][new_y] = '0';
                            }
                        }
                    }
                }
            }
        }
        return island;
    }
}